### ISSUE TITLE
How to use USB Camera in flutter_aws_chime

### ISSUE THREAD

**Issue by rvharjinderbains** [likeconan/flutter_aws_chime#2]
Thank you for the flutter_aws_chime plugin! Currently, it only detects the default device camera. I’d like to request support for external USB cameras, enabling users to select and use them for video streaming.

This feature would be helpful for applications needing specialized or multiple camera setups.

Let me know if more details are needed.

**Comment by likeconan** [comment]
@rvharjinderbains Thanks for your issue, may I ask for USB cameras, what devices are you using? Mobile or desktop?

**Comment by rvharjinderbains** [comment]
Hi @likeconan, Thank you for your prompt response. I am using an Android device and attempting to connect an external camera. To facilitate this connection, I am utilizing a converter from USB to Type-C to connect the USB webcam.

To test the setup, I used the flutter_uvc_camera plugin, which successfully detected the camera on the device. However, I am unable to achieve the same functionality using the flutter_aws_chime plugin.

Please let me know if any additional details are required.

**Comment by rvharjinderbains** [comment]
Hi @likeconan I hope you have checked my last comment. Can you please confirm if you get any solution for that. Or you need any further details

**Comment by rvharjinderbains** [comment]
Hi @likeconan , Could you please confirm if it is possible to connect an external camera to the AWS Chime SDK for mobile applications? If not, would you recommend using an alternative approach, such as utilizing flutter_webrtc or another custom video streaming solution, to handle this requirement?

We are currently at a critical point in our development where integrating an external camera is essential, and this functionality is a key reason for choosing AWS Chime.

I’d appreciate your guidance on this matter. You can also reach me directly via email at harjinder_bains@rvtechnologies.com if needed.

Thank you for your support!

**Comment by likeconan** [comment]
HI @rvharjinderbains, Let me do some research and get back to you. I think it should support external camera, I will give you some guide tomorrow.

**Comment by rvharjinderbains** [comment]
Thanks @likeconan 

Thank you for your response. That sounds great! I’ll wait for your guidance and updates. Let me know if you need any further details or input from my side in the meantime.

Looking forward to hearing from you tomorrow.

**Comment by likeconan** [comment]
@rvharjinderbains Just found two ways to resolve it

1. Go to kotlin code in Android, check below code if it's working

```
val videoDevices = MeetingSessionManager.meetingSession?.audioVideo?.listVideoInputDevices()
val selectedDevice = videoDevices?.get(indexOfDesiredCamera) // try to use the external camera
MeetingSessionManager.meetingSession?.audioVideo?.chooseVideoInputDevice(selectedDevice?.deviceId)
```
You can have a try to put it in below code to have a test
<img width="1544" alt="image" src="https://github.com/user-attachments/assets/a7de904a-3148-4b43-88c6-7c5cb9830f1b" />

If everything works as expected, then you can just use a flutter channel for further customization function.

2. Since it has been a long time for updating, I found that awschimesdk has updated and supported a lot of advanced functions now, as you can see from below:

https://github.com/aws/amazon-chime-sdk-android/blob/master/guides/custom_video.md

<img width="1201" alt="image" src="https://github.com/user-attachments/assets/9a9f3937-dfb4-489c-8bfe-295994508ee1" />

If first solution doers not work, then I need to upgrade my repo to support it and it will take a long time...

Hope this answers your question.
Thanks

**Comment by rvharjinderbains** [comment]
Hi @likeconan 

Thank you for providing the detailed solution for integrating external cameras with the flutter_aws_chime plugin. While implementing the suggested changes, I encountered the following errors in the MethodChannelCoordinator.kt file:

```
Unresolved Reference Errors:
MethodChannelCoordinator.kt: (186, 78): Unresolved reference: listVideoInputDevices
MethodChannelCoordinator.kt: (188, 59): Unresolved reference: chooseVideoInputDevice
```

It seems these methods are not recognized in the current implementation. Additionally, I believe we need to handle some extra permissions to access the USB external camera. Specifically, the following permissions and features appear to be required:

```
<uses-permission android:name="android.permission.USB_PERMISSION" />
<uses-feature android:name="android.hardware.usb.host" />
<uses-feature android:name="android.hardware.camera" />
```


I also came across a GitHub repository, [flutter_uvc_camera](https://github.com/chenyeju295/flutter_uvc_camera/tree/master), which provides an example of handling USB cameras in a Flutter plugin. This repository includes the necessary permissions and logic to interact with USB cameras. It successfully demonstrates the use of external camera views, and I believe incorporating a similar approach into the flutter_aws_chime plugin would resolve this issue.

Could you please review the repository and provide guidance on integrating the required functionality into flutter_aws_chime?

Thank you for your assistance!

**Comment by likeconan** [comment]
@rvharjinderbains Well, even though [flutter_uvc_camera](https://github.com/chenyeju295/flutter_uvc_camera/tree/master) did the trick to get USB camera working, the tricky part is letting AWSChimeSDK to use the correct camera, if there are no implementation for list video inputs and choose the right one. I think the correct way is to upgrade AWSChimeSDK and implement it

**Comment by rvharjinderbains** [comment]
@likeconan  Thanks for the update! I understand the challenge with getting AWSChimeSDK to properly support USB cameras without the ability to list and select video inputs. Upgrading AWSChimeSDK and implementing this feature sounds like the right approach.

As a suggestion for the Flutter Chime SDK, it would be helpful to include functionality that lists all available cameras (both external and default), allowing users to select their preferred option. This could simplify the process for developers and enhance the user experience.

Let me know your thoughts, and I’m happy to assist further if needed.

**Comment by likeconan** [comment]
@rvharjinderbains Yes, totally agreed, and I will update it when I'm free, there is no timeline guarantee for it. Sorry about that

**Comment by rvharjinderbains** [comment]
@likeconan Thanks for letting me know. I completely understand, and I appreciate you taking the time to work on it when you’re able. If there’s anything I can do to help in the meantime, just let me know!

**Comment by rvharjinderbains** [comment]
Hi @likeconan is there any update?

**Comment by likeconan** [comment]
@rvharjinderbains WIP, not finished yet

**Comment by rvharjinderbains** [comment]
Thanks @likeconan for your response and understood that this is WIP, but could you please implement it ASAP? I require USB camera support in my Flutter app, and I'm waiting to launch it. Let me know if you need any details from my side. Thanks!

**Comment by rvharjinderbains** [comment]
Hi @likeconan , just following up on this — do you have any update on the USB camera support? 🙏
This is a key feature for my Flutter app and I’m waiting on it to move forward with the launch.
Please let me know if you need any details or testing support from my side. Thanks!
